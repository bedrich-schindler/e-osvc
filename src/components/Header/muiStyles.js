import { SIDEBAR_WIDTH } from '../Menu/muiStyles';

export default (theme) => ({
  appBar: {
    transition: theme.transitions.create(['width', 'margin'], {
      duration: theme.transitions.duration.leavingScreen,
      easing: theme.transitions.easing.sharp,
    }),
    zIndex: theme.zIndex.drawer + 1,
  },
  appBarWithOpenedMenu: {
    marginLeft: SIDEBAR_WIDTH,
    transition: theme.transitions.create(['width', 'margin'], {
      duration: theme.transitions.duration.enteringScreen,
      easing: theme.transitions.easing.sharp,
    }),
    width: `calc(100% - ${SIDEBAR_WIDTH}px)`,
  },
});
