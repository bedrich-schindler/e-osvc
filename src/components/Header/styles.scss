.accountMenuButton {
  position: absolute;
  right: 12px;
}

.menuButton {
  margin-right: 5px;
  margin-left: 5px;
}

.title {
  max-width: 20rem;
  margin-left: 2.5rem;
  font-weight: 500;
  font-size: 1.25rem;
  transform: scale(1);
  transition: max-width linear 0.1s, transform linear 0.1s;
}

.userName {
  position: relative;
  top: 0.1rem;
  margin-right: 1rem;
  font-weight: 500;
}

@media all and (max-width: 767px) {
  [class*="Header-appBarWithOpenedMenu"] .title {
    max-width: 0;
    overflow: hidden;
    transform: scale(0);
  }
}

@media all and (max-width: 374px) {
  .title {
    display: none;
  }
}

@media all and (max-width: 599px) {
  .userName {
    display: none;
  }
}
