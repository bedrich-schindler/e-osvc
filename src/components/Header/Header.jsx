import classNames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import AppBar from '@material-ui/core/AppBar';
import IconAccountCircle from '@material-ui/icons/AccountCircle';
import IconButton from '@material-ui/core/IconButton';
import IconMenu from '@material-ui/icons/Menu';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import Toolbar from '@material-ui/core/Toolbar';
import { withStyles } from '@material-ui/core/styles';
import { withRouter } from 'react-router-dom';
import routes from '../../routes';
import muiStyles from './muiStyles';
import styles from './styles.scss';

export class Header extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      accountMenuEl: null,
    };

    this.handleOpenAccountMenu = this.handleOpenAccountMenu.bind(this);
    this.handleCloseAccountMenu = this.handleCloseAccountMenu.bind(this);
  }

  handleOpenAccountMenu(e) {
    this.setState({ accountMenuEl: e.currentTarget });
  }

  handleCloseAccountMenu() {
    this.setState({ accountMenuEl: null });
  }

  renderAccountMenu() {
    const {
      history,
      user,
      onLogout,
    } = this.props;
    const { accountMenuEl } = this.state;

    return (
      <div className={styles.accountMenuButton}>
        <span className={styles.userName}>
          {user.name}
        </span>
        <IconButton
          color="inherit"
          onClick={this.handleOpenAccountMenu}
        >
          <IconAccountCircle />
        </IconButton>
        <Menu
          anchorEl={accountMenuEl}
          open={Boolean(accountMenuEl)}
          onClose={this.handleCloseAccountMenu}
        >
          <MenuItem onClick={() => {
            history.push(routes.account.path);
          }}>
            Můj účet
          </MenuItem>
          <MenuItem onClick={onLogout}>
            Odhlásit se
          </MenuItem>
        </Menu>
      </div>
    );
  }

  renderMenu() {
    const {
      isMenuOpened,
      onOpenMenu,
    } = this.props;

    if (isMenuOpened) {
      return null;
    }

    return (
      <div className={styles.menuButton}>
        <IconButton
          color="inherit"
          onClick={onOpenMenu}
        >
          <IconMenu />
        </IconButton>
      </div>
    );
  }

  renderTitle() {
    const { title } = this.props;

    return (
      <span className={styles.title}>
        {title}
      </span>
    );
  }

  render() {
    const {
      classes,
      isMenuOpened,
    } = this.props;

    return (
      <AppBar
        className={classNames(
          classes.appBar,
          {
            [classes.appBarWithOpenedMenu]: isMenuOpened,
          },
        )}
        position="static"
      >
        <Toolbar disableGutters={!isMenuOpened}>
          {this.renderMenu()}
          {this.renderTitle()}
          {this.renderAccountMenu()}
        </Toolbar>
      </AppBar>
    );
  }
}

Header.propTypes = {
  classes: PropTypes.object.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  isMenuOpened: PropTypes.bool.isRequired,
  onLogout: PropTypes.func.isRequired,
  onOpenMenu: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
};

export default withStyles(muiStyles, { withTheme: true })(withRouter(Header));
