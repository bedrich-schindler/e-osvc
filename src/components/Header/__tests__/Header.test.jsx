import React from 'react';
import sinon from 'sinon';
import {
  mount,
  shallow,
} from 'enzyme';
import { Header } from '../Header';

describe('rendering', () => {
  it('renders correctly', () => {
    const tree = shallow(
      <Header
        classes={{}}
        history={{ push: () => { }}}
        title="Title"
        isMenuOpened={false}
        user={{ name: 'Name' }}
        onLogout={() => {}}
        onOpenMenu={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with opened menu', () => {
    const tree = shallow(
      <Header
        classes={{}}
        history={{ push: () => { }}}
        title="Title"
        isMenuOpened
        user={{ name: 'Name' }}
        onLogout={() => {}}
        onOpenMenu={() => {}}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onOpenMenu when menu button is clicked', () => {
    const spy = sinon.spy();
    const tree = mount(
      <Header
        classes={{}}
        history={{ push: () => { }}}
        title="Title"
        isMenuOpened={false}
        user={{ name: 'Name' }}
        onLogout={() => {}}
        onOpenMenu={spy}
      />,
    );

    tree.find('button').at(0).simulate('click');
    expect(spy.calledOnce).toEqual(true);
  });
});
