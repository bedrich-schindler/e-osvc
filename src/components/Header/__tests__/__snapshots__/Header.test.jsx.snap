// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  className=""
  position="static"
>
  <WithStyles(ForwardRef(Toolbar))
    disableGutters={true}
  >
    <div
      className="menuButton"
    >
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        onClick={[Function]}
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
    </div>
    <span
      className="title"
    >
      Title
    </span>
    <div
      className="accountMenuButton"
    >
      <span
        className="userName"
      >
        Name
      </span>
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        onClick={[Function]}
      >
        <AccountCircleIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        onClose={[Function]}
        open={false}
      >
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Odhlásit se
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;

exports[`rendering renders correctly with opened menu 1`] = `
<WithStyles(ForwardRef(AppBar))
  className="undefined"
  position="static"
>
  <WithStyles(ForwardRef(Toolbar))
    disableGutters={false}
  >
    <span
      className="title"
    >
      Title
    </span>
    <div
      className="accountMenuButton"
    >
      <span
        className="userName"
      >
        Name
      </span>
      <WithStyles(ForwardRef(IconButton))
        color="inherit"
        onClick={[Function]}
      >
        <AccountCircleIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Menu))
        anchorEl={null}
        onClose={[Function]}
        open={false}
      >
        <WithStyles(ForwardRef(MenuItem))
          onClick={[Function]}
        >
          Odhlásit se
        </WithStyles(ForwardRef(MenuItem))>
      </WithStyles(ForwardRef(Menu))>
    </div>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
